core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -306,6 +312,98 @@ default Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessi
+                            }
+                            catch (RuntimeException e) {
+                                // getTableMetadata call may fail if table disappeared during listing
+                                Helper.juliLogger.log(Level.WARNING, () -> "Failed to get metadata for table: " + tableName);
+                                tables = List.of();
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        else {
+            Optional<String> schema = prefix.getSchema();
+            materializedViews = getMaterializedViews(session, schema).entrySet().stream()
+                    .collect(toMap(Map.Entry::getKey, entry -> TableCommentMetadata.forTable(entry.getKey(), entry.getValue().getComment())));
+            views = getViews(session, schema).entrySet().stream()
+                    .collect(toMap(Map.Entry::getKey, entry -> TableCommentMetadata.forTable(entry.getKey(), entry.getValue().getComment())));
+            tables = listTables(session, schema).stream()
+                    .filter(tableName -> !materializedViews.containsKey(tableName) && !views.containsKey(tableName))
+                    .map(tableName -> {
+                        if (redirectTable(session, tableName).isPresent()) {
+                            return TableCommentMetadata.forRedirectedTable(tableName);
+                        }
+                        ConnectorTableHandle tableHandle = getTableHandle(session, tableName, Optional.empty(), Optional.empty());
+                        if (tableHandle == null) {
+                            // disappeared during listing
+                            return null;
+                        }
+                        try {
+                            return TableCommentMetadata.forTable(tableName, getTableMetadata(session, tableHandle).getComment());
+                        }
+                        catch (RuntimeException e) {
+                            // getTableMetadata call may fail if table disappeared during listing
+                            Helper.juliLogger.log(Level.WARNING, () -> "Failed to get metadata for table: " + tableName);
+                            return null;
+                        }
+                    })
+                    .filter(Objects::nonNull)
+                    .toList();
+        }
+
+        return Stream.of(
+                        materializedViews.values().stream(),
+                        views.values().stream()
+                                .filter(commentMetadata -> !materializedViews.containsKey(commentMetadata.table())),
+                        tables.stream()
+                                .filter(commentMetadata -> !materializedViews.containsKey(commentMetadata.table()))
+                                .filter(commentMetadata -> !views.containsKey(commentMetadata.table())))
```

Comment:
Do you also need to check the schema here? A table in one schema could have the same name as a MV in a different schema.

Reply:
key is `SchemaTableName`